python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,277 @@
+    from dagster._utils.caching_instance_queryer import CachingInstanceQueryer
+
+
+class TemporalContext(NamedTuple):
+    """TemporalContext represents an effective time, used for business logic, and last_event_id
+    which is used to identify that state of the event log at some point in time. Put another way,
+    the value of a TemporalContext represents a point in time and a snapshot of the event log.
+
+    Effective time: This is the effective time of the computation in terms of business logic,
+    and it impacts the behavior of partioning and partition mapping. For example, if an
+    if you get the "last" partition window of a given partitions definition, it is with
+    respect to the effective time.
+
+    Last event id: Our event log has a monotonic increasing event id. This is used to
+    cursor the event log. This event_id is also propogated to derived tables to indicate.
+    This allows us to query the state of the event log at a given point in time.
+
+    Note that insertion time of the last_event_id is not the same as the effective time.
+
+    A last_event_id of None indicates that the reads will be volatile will immediately
+    reflect any subsequent writes.
+    """
+
+    effective_dt: datetime
+    last_event_id: Optional[int]
+
+
+class _AssetSliceCompatibleSubset(ValidAssetSubset): ...
+
+
+PartitionKey: TypeAlias = Optional[str]
+AssetPartition: TypeAlias = AssetKeyPartitionKey
+
+
+def _slice_from_subset(asset_graph_view: "AssetGraphView", subset: AssetSubset) -> "AssetSlice":
+    valid_subset = subset.as_valid(
+        asset_graph_view.asset_graph.get_partitions_def(subset.asset_key)
+    )
+    return AssetSlice(asset_graph_view, _AssetSliceCompatibleSubset(*valid_subset))
+
+
+class AssetSlice:
+    """An asset slice represents a set of partitions for a given asset key. It is
+    tied to a particular instance of an AssetGraphView, and is read-only.
+
+    With an asset slice you are able to traverse the set of partitions resident
+    in an asset graph at any point in time.
+
+    ```python
+        asset_graph_view_t0 = AssetGraphView.for_test(defs, effective_dt=some_date())
+
+        some_asset_slice = asset_graph_view_to.get_asset_slice(some_asset.key)
+
+        for parent_slice in some_asset_slice.get_parent_slices().values():
+            # do something with the parent slice
+    ```
+
+    AssetSlice is read-only and tied to a specific AssetGraphView. Therefore
+    we can aggressively cached methods and properties. However different methods
+    have different performance characterics so we have the following conventions:
+
+    Naming conventions
+    * Properties guaranteed to be fast.
+    * Methods prefixed with `get_` do some work in-memory but not hugely expensive
+    * Methods prefixed with `compute_` do potentially expensive work, like compute
+    * partition mappings and query the instance.
+    * Methods using "materialize" indicate that they fully materialize partition sets
```

Comment:
"materialize" is a confusing word here -- seems like `materialize_partition_keys` could equivalently be called `compute_partition_keys`.

Reply:
@OwenKephart I agree that is a concern. However in Python land "materialize" is the word I see most often when describing the process of generating a list in memory from a generator. I.e. "you materialize the generator and get a list". So thats why I chose that word. Could be persuaded to go with `compute` or another word, but I think it is a little different.

Reply:
I think "does potentially expensive work, like compute partition mappings and query the instance" applies well enough to what the consequences of calling this function would be (the underlying work done is basically the same thing that's happening when you do a partition mapping), so to avoid expanding the categorization scheme unnecessarily (and the collision with the very asset-focused "materialize" verb) I would definitely prefer `compute`

Reply:
Cool i'm sold @OwenKephart will update